xarray/backends/plugins.py

```diff
@@ -1,31 +1,89 @@
 import inspect
+import itertools
 import typing as T
+import warnings
+from functools import lru_cache
 
-from . import cfgrib_, h5netcdf_, zarr
-
-ENGINES: T.Dict[str, T.Dict[str, T.Any]] = {
-    "h5netcdf": {
-        "open_dataset": h5netcdf_.open_backend_dataset_h5necdf,
-    },
-    "zarr": {
-        "open_dataset": zarr.open_backend_dataset_zarr,
-    },
-    "cfgrib": {
-        "open_dataset": cfgrib_.open_backend_dataset_cfgrib,
-    },
-}
-
-
-for engine in ENGINES.values():
-    if "signature" not in engine:
-        parameters = inspect.signature(engine["open_dataset"]).parameters
-        for name, param in parameters.items():
-            if param.kind in (
-                inspect.Parameter.VAR_KEYWORD,
-                inspect.Parameter.VAR_POSITIONAL,
-            ):
-                raise TypeError(
-                    f'All the parameters in {engine["open_dataset"]!r} signature should be explicit. '
-                    "*args and **kwargs is not supported"
-                )
-        engine["signature"] = set(parameters)
+import entrypoints  # type: ignore
+
+
+class BackendEntrypoint:
+    __slots__ = ("open_dataset", "open_dataset_parameters")
+
+    def __init__(self, open_dataset, open_dataset_parameters=None):
+        self.open_dataset = open_dataset
+        self.open_dataset_parameters = open_dataset_parameters
+
```

Comment:
If we required `open_dataset_parameters` in the constructor, we could just use a `namedtuple` here. If we foresee any subclassing of the `BackendEntrypoint` class, then I suggest we stick with what you have here.